clang/lib/CodeGen/CGGPUBuiltin.cpp

```diff
@@ -199,15 +214,31 @@ RValue CodeGenFunction::EmitAMDGPUDevicePrintfCallExpr(const CallExpr *E) {
     }
 
     llvm::Value *Arg = A.getRValue(*this).getScalarVal();
+    if (isString(A.getType().getTypePtr()) && CGM.getLangOpts().OpenCL)
```

Comment:
The typecast can be inserted later when the arguments are actually processed. At that point, we already know which args are strings because we have parsed the format string.